### PR TITLE
Handle mismatched nodenames in etcd mode IPAM cleanup

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

Looks like our IPAM GC controller has a bug in etcd mode when the name
of the Node in Calico's etcd doesn't match the name of the node in the
Kubernetes API.

It causes the controller to wrongfully release IP addresses that should
not yet be released. This change updates the controller to map nodenames
from Calico node names to Kubernetes node names before doing the
existence check.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Fix IPAM garbage collection in etcd mode on clusters where node name does not match Kubernetes node name. 
```

### CODE REVIEW

File: pkg/controllers/node/node_syncer.go [hunk 1]
```diff
@@ -63,14 +63,14 @@ func (c *NodeController) OnUpdates(updates []bapi.Update) {
 		case bapi.UpdateTypeKVUpdated:
 			n := upd.KVPair.Value.(*apiv3.Node)
 
-			if c.config.SyncLabels {
```
Comment (caseydavenport) [inline pkg/controllers/node/node_syncer.go:66]: This change just makes sure we're always populating the nodemapper map (we're already always running the syncer.)

File: pkg/controllers/node/ipam.go [hunk 2]
```diff
@@ -77,8 +83,15 @@ func (c *NodeController) syncIPAMCleanup() error {
...
-					nodes[val] = []model.AllocationAttribute{}
+				kn := c.kubernetesNodeForCalico(val)
+				if kn == "" {
+					// If we can't determine a corresponding Kubernetes node for this allocation,
+					// we can't go any further. Move on to the next allocation.
+					logrus.WithField("calicoNode", val).Info("Could not find corresponding k8s node for allocation, skip")
+					continue
+				}
+				if _, ok := nodes[kn]; !ok {
+					nodes[kn] = []model.AllocationAttribute{}
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:94]: Not part of this PR, but.... firstSkip is always true - should be set to false below.

File: pkg/controllers/node/ipam.go [hunk 3]
```diff
@@ -50,17 +53,20 @@ func (c *NodeController) syncIPAMCleanup() error {
...
 	for _, kvp := range blocks.KVPairs {
 		b := kvp.Value.(*model.AllocationBlock)
 
 		// Include affinity if it exists. We want to track nodes even
 		// if there are no IPs actually assigned to that node.
 		if b.Affinity != nil {
-			n := strings.TrimPrefix(*b.Affinity, "host:")
-			if _, ok := nodes[n]; !ok {
-				nodes[n] = []model.AllocationAttribute{}
+			n := c.kubernetesNodeForCalico(strings.TrimPrefix(*b.Affinity, "host:"))
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:65]: This method seems to return either k8s node name or a calico node name depending on what's available.

Could we make a little clearer by:
- Keeping this processing unchanged, perhaps just changing nodes -> calicoNodes to make it clearer
- When we loop through allocations we replace the nodeExists() with something like getKubernetesNodeFromCalico() and then we keep all of the k8s node stuff just within that loop.  It also means the call to cleanupNode can be made with the correct node name (which I think with this code it isn't).

Comment (robbrockbank) [reply pkg/controllers/node/ipam.go:65]: ^ note the second bullet.  I think the call to cleanupNode is called with the wrong node name.

Comment (robbrockbank) [reply pkg/controllers/node/ipam.go:65]: Also perhaps a good opportunity to tighten up naming when we use the name node to make it clear whether calico or kubernetes.

File: pkg/controllers/node/ipam.go [hunk 4]
```diff
@@ -96,24 +100,41 @@ func (c *NodeController) syncIPAMCleanup() error {
...
 	// For each node present in IPAM, if it doesn't exist in the Kubernetes API then we
 	// should consider it a candidate for cleanup.
-	for node, allocations := range nodes {
-		// Check if it exists in the Kubernetes API.
-		logc := log.WithField("node", node)
-		if c.nodeExists(node) {
-			logc.Debug("Node still exists, continue")
+	for cnode, allocations := range calicoNodes {
+		// Lookup the corresponding Kubernetes node for each Calico node we found in IPAM.
+		// In KDD mode, these are 1:1. However, in etcd mode its possible that the Calico node has a
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:116]: nit:  1:1 -> identical.

File: pkg/controllers/node/ipam.go [hunk 5]
```diff
@@ -217,24 +239,56 @@ func (c *NodeController) nodeExists(node string) bool {
 	return true
 }
 
+// podExists returns whether the given pod exists in the Kubernetes API and is on the provided Kubernetes node.
+// Note that the "node" parameter is the name of the Kubernetes node in the Kubernetes API.
 func (c *NodeController) podExists(name, ns, node string) bool {
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:244]: nit:  I think this should be podExistsOnNode() rather than podExists()

File: pkg/controllers/node/ipam.go [hunk 6]
```diff
@@ -217,24 +239,56 @@ func (c *NodeController) nodeExists(node string) bool {
...
-	n, err := c.k8sClientset.CoreV1().Pods(ns).Get(name, v1.GetOptions{})
+	p, err := c.k8sClientset.CoreV1().Pods(ns).Get(name, v1.GetOptions{})
 	if err != nil {
 		if errors.IsNotFound(err) {
 			return false
 		}
 		log.WithError(err).Warn("Failed to query pod, assume it exists")
 	}
-	if n.Spec.NodeName != node {
+	if p.Spec.NodeName != node {
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:252]: Prolly just not grokking correctly, but for borrowed IPs won't this be the case and in that case we'll clean up the allocation.

File: pkg/controllers/node/ipam.go [hunk 7]
```diff
@@ -137,7 +158,7 @@ func (c *NodeController) syncIPAMCleanup() error {
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:158]: Shouldn't we be setting canDelete to false at line 135?

File: pkg/controllers/node/ipam.go [hunk 8]
```diff
@@ -137,7 +158,7 @@ func (c *NodeController) syncIPAMCleanup() error {
 
 			// Check to see if the pod still exists. If it does, then we shouldn't clean up
 			// this node, since it might come back online.
-			if pod != "" && c.podExists(pod, ns, node) {
+			if pod != "" && c.podExists(pod, ns, knode) {
```
Comment (robbrockbank) [inline pkg/controllers/node/ipam.go:161]: nit: Also check `knode != ""`

I think it kinda drops out in the podExists check, but I think it's a bit clearer to explicitly handle the "unknown knode" case.

File: pkg/controllers/node/ipam.go (hunk 7 above)
Comment (caseydavenport) [reply pkg/controllers/node/ipam.go:159]: Yep I think so!
